complex: Add wait object type option.

a-ilango opened the pull request:
Allows user to run tests with FI_WAIT_UNSPEC, FI_WAIT_NONE,
FI_WAIT_FD wait objects in EQ and CQ. Support for other wait
objects will be added later.

shefty commented on complex/ft_domain.c:
```diff
@@ -61,7 +61,7 @@ static int ft_open_eq(void)
 
 	memset(&attr, 0, sizeof attr);
-	attr.wait_obj = FI_WAIT_FD;
+	attr.wait_obj = test_info.eq_wait_obj;
```
The code is prepared to handle FI_WAIT_NONE for an EQ.  ft_get_event() needs to change, so that it polls reading the EQ until an event is found.  It currently calls EQ sread, which isn't supported if wait none is specified.

shefty commented on complex/ft_test.c:
```diff
@@ -49,7 +49,7 @@ static int ft_init_xcontrol(struct ft_xcontrol *ctrl)
 	ctrl->credits = FT_DEFAULT_CREDITS;
 	ctrl->max_credits =  FT_DEFAULT_CREDITS;
-	ft_rx.comp_wait = FI_WAIT_NONE;
+	ctrl->comp_wait = test_info.cq_wait_obj;
```
Somewhat similar, if FI_WAIT_NONE is *not* specified, then the code should call fi_cq_sread() instead of polling on fi_cq_read() in ft_comp_rx/tx. This way, we're at least testing the wait object code path.

For FI_WAIT_FD, we may want to go further an actually get the fd and call select/poll.  Though that's more complexity.  For now, I would just add calling sread.

shefty commented:
A couple of comments, but will be a nice addition.

a-ilango commented:
@shefty I have modified the eq and cq read functions to call different functions based on wait obj. For FI_WAIT_FD, the poll option can be added later on as you say.

shefty commented on complex/ft_comp.c:
```diff
@@ -118,53 +118,40 @@ int ft_bind_comp(struct fid_ep *ep, uint64_t flags)
 }
 
-int ft_comp_rx(void)
+int ft_comp_x(struct fid_cq *cq, struct ft_xcontrol *ft_x, char *x_str, int nowait)
```
A couple of minor requests:  using static char for x_str, and pass in the wait time directly, rather than using a boolean.

a-ilango replied:
Why do we need static char for x_str?

shefty replied:
technically it should be 'static char *'.  The function doesn't change the string.

bturrubiates replied:
@shefty Do you mean `const`?

shefty replied:
Uhm... yeah, that's what I said.  Right?  :P

a-ilango replied:
OK. :) will make it const.

shefty commented on complex/fabtest.h:
```diff
@@ -202,6 +209,7 @@ ssize_t ft_get_event(uint32_t *event, void *buf, size_t len,
 		     uint32_t event_check, size_t len_check);
 int ft_open_comp();
 int ft_bind_comp(struct fid_ep *ep, uint64_t flags);
+int ft_comp_x(struct fid_cq *cq, struct ft_xcontrol *ft_x, char *x_str, int nowait);
```
I would keep ft_comp_x as a static call internal to ft_comp.c, and let the external users call the base ft_comp_rx/tx calls only.  This will simplify changes when we introduce counters.

a-ilango commented:
I have updated the PR with the following changes:
* Added const qualifier for string passed to ft_comp_x and *q_readerr functions.
* Refactored ft_comp_x to include a polling option when there is no wait object.

shefty commented on complex/ft_comp.c:
```diff
@@ -118,53 +118,71 @@ int ft_bind_comp(struct fid_ep *ep, uint64_t flags)
+		break;
+	case FI_WAIT_UNSPEC:
+		ft_cq_read(fi_cq_sread, cq, buf, comp_entry_cnt[ft_x->cq_format],
+			ft_x->credits, x_str, ret, NULL, timeout);
```
I expected this to call fi_cq_sread().

a-ilango replied:
It *is* calling fi_cq_sread. I'm passing the function, fi_cq_sread as argument to the macro, ft_cq_read.

shefty replied:
Ah - I completely missed that.

shefty commented on complex/ft_comp.c:
```diff
@@ -118,53 +118,71 @@ int ft_bind_comp(struct fid_ep *ep, uint64_t flags)
+	case FI_WAIT_FD:
+	case FI_WAIT_SET:
+	case FI_WAIT_MUTEX_COND:
+		FT_ERR("fi_ubertest: Unsupported cq wait object\n");
```
The FD and mutex cases are probably safe to call fi_cq_sread() as well.  We're trusting that the provider has allocated the correct wait object, but it should work for now.  Wait set support would require additional work in the test to allocate the wait set.

a-ilango replied:
OK, I shall update it to call fi_cq_sread for FD and mutex cases.

shefty commented on complex/ft_domain.c:
```diff
@@ -90,12 +90,35 @@ ssize_t ft_get_event(uint32_t *event, void *buf, size_t len,
+		break;
+	case FI_WAIT_FD:
+	case FI_WAIT_SET:
+	case FI_WAIT_MUTEX_COND:
```
Similar comment as the CQ case above.

shefty commented:
It occurs to me that a provider may not support a requested wait object.  And the test won't know this until at run time when it tries to create the CQ. I'm not sure how to resolve this. Ideally, the test doesn't generate an error and successfully aborts as not supported.  Did you try testing this case?

shefty commented:
I guess since the user can control what gets tested through the config file, this isn't a huge deal.  Not enough to block merging this patch.

a-ilango commented:
I just ran the test with a wait object not supported by the provider. The test fails at fi_cq_open on the client side. But the server is not aware of this and the test just hangs. We should probably make server and client abort gracefully instead of a hang.

a-ilango commented:
Updated code to use fi_cq_sread/fi_eq_sread for FI_WAIT_FD and FI_WAIT_MUTEX_COND wait objects.

shefty commented:
Thanks - merging this.  We can provide fix-ups in a follow on patch.  At least there's a work-around for handling options not supported by a provider.
